File: examples/csharp/Heap1.cs
using System;
using System.Collections.Generic;

class Heap1
{
    static List<int> heap = new List<int>();

    static void Main()
    {
        Insert(5);
        Insert(3);
        Insert(8);
        Insert(1);
        Console.WriteLine(ExtractMin());
        Console.WriteLine(ExtractMin());
    }

    static void Insert(int val)
    {
        heap.Add(val);
        int i = heap.Count - 1;
        while (i > 0 && heap[(i - 1) / 2] > heap[i])
        {
            (heap[i], heap[(i - 1) / 2]) = (heap[(i - 1) / 2], heap[i]);
            i = (i - 1) / 2;
        }
    }

    static int ExtractMin()
    {
        if (heap.Count == 0) return -1;
        int min = heap[0];
        heap[0] = heap[^1];
        heap.RemoveAt(heap.Count - 1);
        int i = 0;
        while (true)
        {
            int smallest = i, left = 2 * i + 1, right = 2 * i + 2;
            if (left < heap.Count && heap[left] < heap[smallest]) smallest = left;
            if (right < heap.Count && heap[right] < heap[smallest]) smallest = right;
            if (smallest == i) break;
            (heap[i], heap[smallest]) = (heap[smallest], heap[i]);
            i = smallest;
        }
        return min;
    }
}

File: examples/java/Heap1.java
import java.util.ArrayList;

public class Heap1 {
    static ArrayList<Integer> heap = new ArrayList<>();

    static void swap(int i, int j) {
        int t = heap.get(i);
        heap.set(i, heap.get(j));
        heap.set(j, t);
    }

    static void insert(int x) {
        heap.add(x);
        int i = heap.size() - 1;
        while (i > 0 && heap.get((i - 1) / 2) > heap.get(i)) {
            swap(i, (i - 1) / 2);
            i = (i - 1) / 2;
        }
    }

    static int extractMin() {
        if (heap.isEmpty()) return -1;
        int min = heap.get(0);
        heap.set(0, heap.get(heap.size() - 1));
        heap.remove(heap.size() - 1);
        int i = 0, n = heap.size();
        while (2 * i + 1 < n) {
            int j = 2 * i + 1;
            if (j + 1 < n && heap.get(j + 1) < heap.get(j)) j++;
            if (heap.get(i) <= heap.get(j)) break;
            swap(i, j);
            i = j;
        }
        return min;
    }

    public static void main(String[] args) {
        insert(30);
        insert(10);
        insert(20);
        insert(5);
        System.out.println(extractMin());
        System.out.println(extractMin());
    }
}
